### PR TITLE
[serve] recursive cancellation

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

Implement full recursive cancellation at Serve layer.

Currently, at the time that a parent request is cancelled, if:
- The child request has not been assigned to a replica, _meaning_ the replica scheduler in the router is still processing that request
- AND the parent request is not directly blocking and waiting on the result of the child request

Then the child request will not be correctly cancelled. Note that if the second bullet point were actually flipped, i.e. the parent request _was_ waiting on the result at the time of cancellation, then an `asyncio.CancelledError` will be sent directly to the replica scheduler and the child request will get successfully cancelled. However obviously this cannot be guaranteed to be true.

This PR adds a cache for asyncio tasks that are launched in the router to assign a request to a replica. When a parent request is cancelled, if there are any "pending assignment requests" in the cache, they are cancelled.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/tests/test_cancellation.py [hunk 1]
```diff
@@ -365,5 +365,104 @@ async def get_out_of_band_response(self):
     assert h.get_out_of_band_response.remote().result() == "ok"
 
 
+def test_recursive_cancellation_during_execution(serve_instance):
```
Comment (edoakes) [inline python/ray/serve/tests/test_cancellation.py:368]: in order for these tests to be meaningful, don't we need to turn off the core cancellation support?

Comment (zcin) [reply python/ray/serve/tests/test_cancellation.py:368]: So at first I took the approach of implementing full recursive cancellation at the Serve layer, but that leads to more of a performance drop. I think it's unnecessary to cover the part that core already covers (recursive task cancellation). We just need to take care of canceling the asyncio request assignment task before a request has been assigned to a replica.

So in this test file `test_recursive_cancellation_during_execution` already passes, and `test_recursive_cancellation_during_assignment` is the one that doesn't pass without this PR.

File: python/ray/serve/_private/replica.py [hunk 2]
```diff
@@ -359,6 +359,15 @@ def _wrap_user_method_call(self, request_metadata: RequestMetadata):
...
         except asyncio.CancelledError as e:
             user_exception = e
+
+            # Recursively cancel child requests
+            requests_pending_assignment = (
+                ray.serve.context._get_requests_pending_assignment(
+                    request_metadata.request_id
+                )
+            )
+            for t in list(requests_pending_assignment.values()):
```
Comment (GeneDer) [inline python/ray/serve/_private/replica.py:369]: Nit: Maybe spell out `task` so it's more understanding what this is iterating over 🙃 

Also, we don't need to cast it to a list before iterate through it right?

File: python/ray/serve/_private/router.py [hunk 3]
```diff
@@ -594,6 +599,17 @@ async def assign_request(
     ) -> ReplicaResult:
         """Assign a request to a replica and return the resulting object_ref."""
 
+        response_id = uuid.uuid4()
```
Comment (GeneDer) [inline python/ray/serve/_private/router.py:602]: Non-blocker: There is this [generate_request_id()](https://github.com/ray-project/ray/blob/671844571defbfc0c1349f3732e6233bf685283e/python/ray/serve/_private/utils.py#L552) maybe we can rename and reuse here?

File: python/ray/serve/context.py [hunk 4]
```diff
@@ -205,3 +207,23 @@ def _set_request_context(
...
         )
     )
+
+
+# A map from current request ID to a set of ReplicaResults corresponding
+# to the requests the replica has sent during the current request
+_requests_pending_assignment: Dict[str, Dict[str, asyncio.Task]] = defaultdict(dict)
+
+
+def _get_requests_pending_assignment(parent_request_id: str) -> Dict:
```
Comment (GeneDer) [inline python/ray/serve/context.py:217]: nit
```suggestion
def _get_requests_pending_assignment(parent_request_id: str) -> Dict[str, asyncio.Task]:
```

File: python/ray/serve/context.py [hunk 5]
```diff
@@ -205,3 +207,23 @@ def _set_request_context(
...
+    if parent_request_id in _requests_pending_assignment:
+        return _requests_pending_assignment[parent_request_id]
+
```
Comment (GeneDer) [inline python/ray/serve/context.py:220]: Maybe we can return empty dictionary for the else case?
```suggestion
    return {}
```

File: python/ray/serve/_private/router.py [hunk 6]
```diff
@@ -594,6 +599,17 @@ async def assign_request(
...
+        assign_request_task = asyncio.current_task()
+        ray.serve.context._add_request_pending_assignment(
+            request_meta.request_id, response_id, assign_request_task
```
Comment (GeneDer) [inline python/ray/serve/_private/router.py:605]: We should probably use `internal_request_id` for this. `request_id` can be set by the client from the header and can be repeated/ other unexpected behavior.

File: python/ray/serve/_private/router.py [hunk 7]
```diff
@@ -594,6 +599,17 @@ async def assign_request(
...
+        )
+        assign_request_task.add_done_callback(
+            lambda _: ray.serve.context._remove_request_pending_assignment(
+                request_meta.request_id, response_id
```
Comment (GeneDer) [inline python/ray/serve/_private/router.py:609]: Same for this one

File: python/ray/serve/context.py [hunk 8]
```diff
@@ -205,3 +207,23 @@ def _set_request_context(
...
+
+def _add_request_pending_assignment(parent_request_id: str, response_id: str, task):
+    if parent_request_id:
```
Comment (GeneDer) [inline python/ray/serve/context.py:223]: This if statement will probably always be true? (if we switch to use `internal_request_id`) :)

File: python/ray/serve/context.py [hunk 9]
```diff
@@ -205,3 +207,23 @@ def _set_request_context(
...
+        _requests_pending_assignment[parent_request_id][response_id] = task
+
+
+def _remove_request_pending_assignment(parent_request_id: str, response_id: str):
+    if response_id in _requests_pending_assignment[parent_request_id]:
+        del _requests_pending_assignment[parent_request_id][response_id]
```
Comment (GeneDer) [inline python/ray/serve/context.py:229]: We should probably also call `del _requests_pending_assignment[parent_request_id]` when it's empty else we will keep accumulating request keys with empty dict.

File: python/ray/serve/_private/router.py (hunk 6 above)
Comment (zcin) [reply python/ray/serve/_private/router.py:605]: nice thanks for the suggestion, changed to use internal request id!

**Review by edoakes — APPROVED** [review]

**Review by GeneDer — APPROVED** [review]
:shipit:

**Review by GeneDer — APPROVED** [review]
Thanks Cindy!

**Pull request merged**